File: backtrack/array_sum_combinations/array_sum_combinations.go
package main

import (
	"fmt"
)

// this could be lazy with goroutines
func product(arrs ...[]int) [][]int {
	ans := [][]int{[]int{}}
	for _, arr := range arrs {
		tmp := [][]int{}
		for _, x := range ans {
			for _, y := range arr {
				tmp = append(tmp, append(x, y))
			}
		}
		ans = tmp
	}
	return ans
}

func counter(iterable []int) map[int]int {
	m := make(map[int]int)
	for _, v := range iterable {
		m[v] = m[v] + 1
	}
	return m
}

func sum(iterable []int) int {
	ans := 0
	for _, v := range iterable {
		ans += v
	}
	return ans
}

func sumCombinations(target int, arrs ...[]int) [][]int {
	rest := arrs[0 : len(arrs)-1]
	last := arrs[len(arrs)-1]
	counterLast := counter(last)
	ans := [][]int{}
	for _, combination := range product(rest...) {
		difference := target - sum(combination)
		times, ok := counterLast[difference]
		if ok {
			for i := 0; i < times; i++ {
				ans = append(ans, append(combination, difference))
			}
		}
	}
	return ans
}

func main() {
	A := []int{1, 2, 3, 3}
	B := []int{2, 3, 3, 4}
	C := []int{1, 2, 2, 2}
	target := 7
	result := sumCombinations(target, A, B, C)
	fmt.Println(len(result))
	fmt.Println(result)
}

File: backtrack/array_sum_combinations/array_sum_combinations.py
from collections import Counter
from itertools import product


# Can be improved via FFT
# https://cstheory.stackexchange.com/questions/32036/finding-witness-in-minkowski-sum-of-integers


def sum_combinations(target, *arrs):
    counter_last = Counter(arrs[-1])
    for combination in product(*arrs[:-1]):
        difference = target - sum(combination)
        if difference in counter_last:
            for repeated in range(counter_last[difference]):
                yield combination + (difference,)


A = [1, 2, 3, 3]
B = [2, 3, 3, 4]
C = [1, 2, 2, 2]
target = 7

solution = list(sum_combinations(target, A, B, C))

print(len(solution))
for combination in solution:
    print(combination)
